python/dgl/heterograph.py

```diff
@@ -0,0 +1,1214 @@
+    node_types : list
+        The node types.
+    edge_connections_by_type : dict
+        Specifies how edges would connect nodes of the source type to nodes of
+        the destination type in the following form:
+
+            {edge_type: (source_node_id_tensor, destination_node_id_tensor)}
+
+        where
+
+        * ``source_node_id_tensor`` and ``destination_node_id_tensor`` are
+          IDs within the source and destination node type respectively.
+        * ``edge_type`` is a triplet of
+
+              (source_node_type_name,
+               destination_node_type_name,
+               edge_type_name)
+
+    Examples
+    --------
+    Suppose that we want to construct the following heterogeneous graph:
+
+    .. graphviz::
+
+       digraph G {
+           Alice -> Bob [label=follows]
+           Bob -> Carol [label=follows]
+           Alice -> Tetris [label=plays]
+           Bob -> Tetris [label=plays]
+           Bob -> Minecraft [label=plays]
+           Carol -> Minecraft [label=plays]
+           Nintendo -> Tetris [label=develops]
+           Mojang -> Minecraft [label=develops]
+           {rank=source; Alice; Bob; Carol}
+           {rank=sink; Nintendo; Mojang}
+       }
+
+    One can analyze the graph and figure out the metagraph as follows:
+
+    .. graphviz::
+
+       digraph G {
+           User -> User [label=follows]
+           User -> Game [label=plays]
+           Developer -> Game [label=develops]
+       }
+
+    Suppose that one maps the users, games and developers to the following
+    IDs:
+
+        User name   Alice   Bob     Carol
+        User ID     0       1       2
+
+        Game name   Tetris  Minecraft
+        Game ID     0       1
+
+        Developer name  Nintendo    Mojang
+        Developer ID    0           1
+
+    One can construct the graph as follows:
+
+    >>> import networkx as nx
+    >>> metagraph = nx.MultiGraph([
+    ...     ('user', 'user', 'follows'),
+    ...     ('user', 'game', 'plays'),
+    ...     ('developer', 'game', 'develops')])
+    >>> g = DGLBaseHeteroGraph(
+    ...     metagraph=metagraph,
+    ...     node_types=['user'] * 4 + ['game'] * 2 + ['developer'] * 2,
+    ...     edge_connections_by_type={
+    ...         # Alice follows Bob and Bob follows Carol
+    ...         ('user', 'user', 'follows'): ([0, 1], [1, 2]),
+    ...         # Alice and Bob play Tetris and Bob and Carol play Minecraft
+    ...         ('user', 'game', 'plays'): ([0, 1, 1, 2], [0, 0, 1, 1]),
+    ...         # Nintendo develops Tetris and Mojang develops Minecraft
+    ...         ('developer', 'game', 'develops'): ([0, 1], [0, 1])})
+    """
+
+    # pylint: disable=unused-argument
+    def __init__(
+            self,
+            metagraph,
+            node_types,
+            edge_connections_by_type):
+        super(DGLBaseHeteroGraph, self).__init__()
+
+    # TODO: should this return DGLBaseHeteroGraph or something like
+    # DGLBaseHeteroSubgraph?
```

Comment:
i think it should be a subgraph type. otherwise, it's weird that the updates on the node data of this graph also reflect on the nodes of the parent graph.

Reply:
BTW, if the heterogeneous graph is mutable graph, getting a subgraph is going to be expensive. nodes of the same type are no longer assigned to a single Id range. slicing requires id remapping.

Reply:
Re subgraph: adopted
Re mutable: In the case of mutable heterogeneous graph, I guess we have to make a `node_type x number_of_nodes x number_of_successors` adjacency list.  For now we can focus on implementing this interface for immutable graphs only since that's more important.

Reply:
what is `node_type x number_of_nodes x number_of_successors` adjacency list?

Reply:
`std::vector_by_type< std::vector_by_id< std::pair<type, id>>>`

Reply:
i see. I guess having the data structure won't reduce the overhead of constructing a subgraph. This is my concern. Extensively using the subgraph API for computation might make the computation very slow.

Reply:
For computation like `g['user'].update_all(...)` I think we can translate the call to `g.update_all('user', ...)` *without* materializing the subgraph (i.e. no remapping stuff).  I updated the return type to `DGLBaseHeteroGraphView` to demonstrate my idea.

Reply:
does it mean DGLHeteroGraph also supports `update_all('user', ...)`?

Reply:
No.  For `update_all()` the graph must have one node type and one edge type (see docstring).

Reply:
do we allow `update_all` on a bipartite graph?

Reply:
I believe `g['user', 'item', 'purchase'].send_and_recv(ALL)` should suffice for at least GCMC-like computations: this computes the message from each user to each of his/her purchased item, and the items aggregate the messages from the users and update themselves.